ff/serving/provider/offline.go

```diff
@@ -89,13 +170,177 @@ func (store *memoryOfflineStore) CreateResourceTable(id ResourceID) (OfflineTabl
+	if id.Type != Feature {
+		return nil, errors.New("Only features can be materialized")
+	}
+	table, err := store.getMemoryResourceTable(id)
+	if err != nil {
+		return nil, err
+	}
+	matData := make(materializedRecords, 0, len(table.entityMap))
+	for _, records := range table.entityMap {
+		matRec := latestRecord(records)
+		matData = append(matData, matRec)
+	}
+	sort.Sort(matData)
+	matId := MaterializationID(uuid.NewString())
+	mat := &memoryMaterialization{
+		id:   matId,
+		data: matData,
+	}
+	store.materializations[matId] = mat
+	return mat, nil
+}
+
+type MaterializationNotFound struct {
+	id MaterializationID
+}
+
+func (err *MaterializationNotFound) Error() string {
+	return fmt.Sprintf("Materialization %s not found", err.id)
+}
+
+func (store *memoryOfflineStore) GetMaterialization(id MaterializationID) (Materialization, error) {
+	mat, has := store.materializations[id]
+	if !has {
+		return nil, &MaterializationNotFound{id}
+	}
+	return mat, nil
+}
+
+func latestRecord(recs []ResourceRecord) ResourceRecord {
+	latest := recs[0]
+	for _, rec := range recs {
+		if latest.TS.Before(rec.TS) {
+			latest = rec
+		}
+	}
+	return latest
+}
+
+func (store *memoryOfflineStore) CreateTrainingSet(def TrainingSetDef) error {
+	if err := def.check(); err != nil {
+		return err
+	}
+	label, err := store.getMemoryResourceTable(def.Label)
+	if err != nil {
+		return err
+	}
+	features := make([]*memoryOfflineTable, len(def.Features))
+	for i, id := range def.Features {
+		feature, err := store.getMemoryResourceTable(id)
+		if err != nil {
+			return err
+		}
+		features[i] = feature
+	}
+	labelRecs := label.records()
+	trainingData := make([]trainingRow, len(labelRecs))
+	for i, rec := range labelRecs {
+		featureVals := make([]interface{}, len(features))
+		for i, feature := range features {
+			featureVals[i] = feature.getLastValueBefore(rec.Entity, rec.TS)
+		}
+		labelVal := rec.Value
+		trainingData[i] = trainingRow{
+			Features: featureVals,
+			Label:    labelVal,
+		}
+	}
+	store.trainingSets[def.ID] = trainingData
+	return nil
+}
+
+func (store *memoryOfflineStore) GetTrainingSet(id ResourceID) (TrainingSetIterator, error) {
+	if err := id.check(TrainingSet); err != nil {
+		return nil, err
+	}
+	data, has := store.trainingSets[id]
+	if !has {
+		return nil, &TrainingSetNotFound{id}
+	}
+	return data.Iterator(), nil
+}
+
+type TrainingSetNotFound struct {
+	ID ResourceID
+}
+
+func (err *TrainingSetNotFound) Error() string {
+	return fmt.Sprintf("TrainingSet with ID %v not found", err.ID)
+}
+
+type trainingRows []trainingRow
+
+func (rows trainingRows) Iterator() TrainingSetIterator {
+	return newTrainingRowsIterator(rows)
+}
+
+type trainingRow struct {
+	Features []interface{}
+	Label    interface{}
+}
+
+type trainingRowsIterator struct {
+	data trainingRows
+	idx  int
+}
+
+func newTrainingRowsIterator(data trainingRows) TrainingSetIterator {
+	return &trainingRowsIterator{
```

Comment:
perhaps call this memoryTrainingSetIterator to be more specific on the implementation

Reply:
good call